warehouse/macaroons/services.py

```diff
@@ -0,0 +1,81 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pymacaroons
+
+from sqlalchemy.orm import joinedload
+from sqlalchemy.orm.exc import NoResultFound
+from zope.interface import implementer
+
+from warehouse.accounts.models import User
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.macaroons.models import Macaroon
+
+
+class InvalidMacaroon(Exception):
+    ...
+
+
+@implementer(IMacaroonService)
+class DatabaseMacaroonService:
+    def __init__(self, db_session):
+        self.db = db_session
+
+    def _get_record_from_db(self, macaroon):
+        # TODO: Better handle parsing the identifier here.
+        macaroon_id = macaroon.identifier.split()[1].split(b":")[1].decode("utf8")
+
+        try:
+            dm = (
+                self.db.query(Macaroon)
+                .options(joinedload("user"))
+                .filter(Macaroon.id == macaroon_id)
+                .one()
+            )
+        except NoResultFound:
+            return
```

Comment:
I don't really like using this technique of returning a special value in Python. It makes the function return different data types in different scenarios.
It's basically what we're forced to do in C or Go: requiring the user of this function to not forget to check the return value.
I think it's better to communicate errors via exceptions.

Reply:
This is consistent with the account service methods, which also return `None` for most operations. The exception is verification methods, which (generally) raise to indicate a failure. I don't have a strong opinion here, but I think it's fine in the context: these methods get called within a larger set of checks, whose failure as a whole is usually reported via either an exception or some dedicated type like `Denied`.